src/secrets/src/lib.rs

```diff
@@ -58,25 +63,48 @@ pub struct SecretsReaderConfig {
 /// Securely reads secrets that are managed by a [`SecretsController`].
 ///
 /// Does not provide access to create, update, or delete the secrets within.
+#[derive(Debug, Clone)]
 pub struct SecretsReader {
-    config: SecretsReaderConfig,
+    config: Arc<SecretsReaderConfig>,
 }
 
 impl SecretsReader {
     pub fn new(config: SecretsReaderConfig) -> Self {
-        Self { config }
+        Self {
+            config: Arc::new(config),
+        }
     }
 
     /// Returns the contents of a secret identified by GlobalId
+    ///
+    /// This `read` will return a complete version of the secret even though `SecretsReader` is `Clone` and we could
+    /// have concurrent reads.  It will not return, e.g., one block from v1 and another from v2.
+    ///
+    /// - On Linux / OSX filesystems, `File::open` will hold a handle open so that even if the file is deleted, we still
+    ///   continue to read from it.  This means a SecretOp::Delete followed by a SecretOp::Ensure can never "swap out"
+    ///   data mid-`read_to_end`.
+    /// - We do not allow editing secrets which _would_ expose us to this issue.
+    ///
+    /// (N.B. Were we ever to run with Windows / NTFS, this would also work properly _and_ mid-read edits would be
+    /// disallowed)
```

Comment:
I'm confused! What does `SecretsReader` being `Clone` have to do with concurrent reads? And why would concurrent reads be a problem? Something about a hypothetical alternative design of this reader in which it stores a file descriptor rather than a path? What's the rationale behind inlining `fs::read`? Protection against some hypothetical adversarial stdlib implementation that closes and reopens repeatedly while reading?

Was there something wrong with the original code or commentary? ISTM like the original code was correct and worked as described—but maybe I'm missing something!

Reply:
If `SecretsReader` is `Clone`, then we can have concurrent reads from disk.

But you're right that the issue I describe isn't really related to reads being concurrent -- the comment is really just clarifying _why_ the current implementation of reading from disk is safe.  (You could imagine, e.g., a silly impl that, if it hits an error, it tries to be helpful and reopens the file and continues reading from where it left off or something)

The old code was correct but I inlined the impl because 1) it's short and 2) it makes it clear the same file handle is used the whole time.  But if that just makes things more confusing, I can "out-line" it

Reply:
I think my confusion stems mostly from the comment in the middle, which implies that there is something wrong with `fs::read` that inlining its implementation fixes. I left you a rewording suggestion below that avoids bringing `fs::read` into the mix.